Feature/profile vm2

OlegAndrus12 opened the pull request:
Autorized lecturer and student are tested.
We ought to change design, cause profile page looks very poor for lecturer.
Sould we add service for getting lecturer knowing mark

OlegAndrus12 committed:
Base ViewModel for profile page

OlegAndrus12 committed:
Marks Table almost done

OlegAndrus12 committed:
Pagination in MarksTabe done

OlegAndrus12 committed:
Humanize Lecturer and marks table fixes

Do you know that in c# 1/2 = 0?
Where is my Python?(

OlegAndrus12 committed:
Bug with lector log in fixed

RomanLevkovych commented on DataServices/SqliteDbContext.cs:
```diff
@@ -16,7 +16,7 @@ public SqliteDbContext(string connectionString = null)
             var connectionStringBuilder = new SqliteConnectionStringBuilder();
             // TODO: create adequate location for DB.
-            connectionStringBuilder.DataSource = @"..\..\..\..\DataServices\univerity-db.db";
+            connectionStringBuilder.DataSource = @"C:\Users\andru\Documents\GitHub\Lohyna-project\DataServices\univerity-db.db";
```
Remove this from commit, please

RomanLevkovych commented on View/Pages/ProfilePage.xaml.cs:
```diff
@@ -20,14 +24,53 @@ namespace View.Pages
         public ProfilePage()
+
         {
+            this.logic = new ProfilePageVM(new SQLiteDataService(), "LubomyrHalamaga");
```
why user is hardcoded?

RomanLevkovych commented on View/Pages/ProfilePage.xaml.cs:
```diff
@@ -49,5 +92,70 @@ private void Navbar_Button_Profile_Click(object sender, RoutedEventArgs e)
+            }
+        }
+
+        private void Content_Button_prev_MouseDown(object sender, MouseButtonEventArgs e)
```
Please, leave meaningful names for your methods, not generated one

RomanLevkovych commented on View/Widgets/ProfilePageContent.xaml:
```diff
@@ -22,22 +22,22 @@
+        <TextBlock x:Name="name_surname_textblock" Grid.Row="1" Grid.Column="2" Grid.ColumnSpan="2" Text="Roman Levkovych" FontSize="30" HorizontalAlignment="Center" VerticalAlignment="Center" TextWrapping="WrapWithOverflow"/>
+        <TextBlock x:Name="label2" Text="Course : " Grid.Row="3" Grid.Column="2" VerticalAlignment="Center" FontSize="22" Margin="10 0 0 0"></TextBlock>
+        <TextBlock x:Name="group_department_textblock"  Grid.Row="2" Text="PMi-33" Grid.Column="3" FontSize="22" VerticalAlignment="Center"  Margin="10 0 0 0" TextWrapping="WrapWithOverflow"/>
+        <TextBlock x:Name="label3" Text="Nom zal : " Grid.Row="4" Grid.Column="2" VerticalAlignment="Center" FontSize="22" Margin="10 0 0 0"/>
```
The same for names. If you youse a name - give it a meaningful name, otherwise - remove it

RomanLevkovych commented on View/Widgets/ProfilePageTableMarksHead.xaml:
```diff
@@ -13,9 +13,9 @@
+        <Button x:Name="subject_column" Grid.Column="0" Content="Subject" FontSize="20" Cursor="Hand" Style="{StaticResource HeaderRowButton}"/>
+        <Button x:Name="mark_column" Grid.Column="1" Content="Mark" FontSize="20" Cursor="Hand" Style="{StaticResource HeaderRowButton}"/>
+        <Button x:Name="lecturer_column" Grid.Column="2" Content="Lecturer" FontSize="20" Cursor="Hand" Style="{StaticResource HeaderRowButton}"/>
+        <Button x:Name="semester_column" Grid.Column="3" Content="Semester" FontSize="20" Cursor="Hand" Style="{StaticResource HeaderRowButton}"/>
```
No underscores as word-separator, only C(c)amelCase.

RomanLevkovych commented on ViewModel/ProfilePageVM.cs:
```diff
@@ -1,10 +1,85 @@
+                currentPageMarks = marks.GetRange(current_page_number * page_limit, (current_page_number + 1) * page_limit - current_page_number * page_limit);
+            if (current_page_number == page_count-1 && marks.Count % page_count != 0)
+                currentPageMarks = marks.GetRange(current_page_number * page_limit, marks.Count % page_limit);
+            return currentPageMarks;
```
camelCase

RomanLevkovych commented on ViewModel/ProfilePageVM.cs:
```diff
@@ -1,10 +1,85 @@
+        private List<Rating> marks;
+        private List<Rating> currentPageMarks;
+
+        public ProfilePageVM(SQLiteDataService dataService, string username)
```
You have to use appropriate service in ctor of VM

RomanLevkovych replied:
It hasn't been fixed yet.

RomanLevkovych commented on ViewModel/ProfilePageVM.cs:
```diff
@@ -1,10 +1,85 @@
+            this.personService = dataService;
+            this.groupService = dataService;
+            this.markService = dataService;
+            this.username = username;
```
`this` is redundant here.

OlegAndrus12 committed:
Code refactoring made

Ratings table changed.
It`s working)

OlegAndrus12 committed:
camelCase refactoring

OlegAndrus12 committed:
Bug fixed

OlegAndrus12 committed:
Mocked data deleted and db filled

OlegAndrus12 committed:
Almost done

Fellas, we shoud find the way to work with path correctly

RomanLevkovych commented on View/Pages/ProfilePage.xaml.cs:
```diff
@@ -20,34 +24,154 @@ namespace View.Pages
+                BitmapImage temp = new BitmapImage();
+                temp.BeginInit();
+                temp.CacheOption = BitmapCacheOption.OnLoad;
+                temp.UriSource = new Uri("C:/Users/andru/Documents/GitHub/Lohyna-project/View/Widgets/Images/profile_placeholder.jpg", UriKind.Relative);
```
Don't use hardware or software specific data/calls. Use standard language structures for this.

RomanLevkovych commented on View/Pages/ProfilePage.xaml.cs:
```diff
@@ -20,34 +24,154 @@ namespace View.Pages
+                if (logic.GetLecturer() != null)
+                {
+                    content.groupDepartmentLabel.Text = "Department:";
+                    content.groupDepartmentTextblock.Text = $"{logic.GetLecturer().Department}";
```
Do we really need to wrap it in a string literal?

OlegAndrus12 replied:
Does it really matters?

RomanLevkovych replied:
Yes

RomanLevkovych commented on View/Pages/ProfilePage.xaml.cs:
```diff
@@ -20,34 +24,154 @@ namespace View.Pages
+                    logic.GetRatings();
+                    content.groupDepartmentTextblock.Text = $"{logic.GetStudent().GroupID}";
+                    content.courseTextblock.Text = $"{logic.GetGroup().Course}";
+                    content.numberZalikovkaTextblock.Text = $"{logic.GetStudent().TicketNumber}";
```
The same

RomanLevkovych commented on View/Pages/ProfilePage.xaml.cs:
```diff
@@ -20,34 +24,154 @@ namespace View.Pages
+                if (FillTableIndex % 2 == 0)
+                    this.content.marks.stack1.Children.Add(item);
+                else
+                    this.content.marks.stack2.Children.Add(item);
```
`this` is redundant here, remove it and prefere `var` over type specification.

RomanLevkovych commented on View/Pages/ProfilePage.xaml.cs:
```diff
@@ -20,34 +24,154 @@ namespace View.Pages
             this.NavigationService.Navigate(new Uri("Pages/ProfilePage.xaml", UriKind.Relative));
         }
+
+        private void FillMarksTable()
```
Maybe, we should change the signature of this method to take a data for the table?

RomanLevkovych commented on View/Widgets/ProfilePageTableMarksRow.xaml.cs:
```diff
@@ -20,9 +21,18 @@ namespace View.Widgets
+
+        private void markTextbox_TextChanged(object sender, TextChangedEventArgs e)
+        {
+
```
Remove the empty method

RomanLevkovych commented on ViewModel/ProfilePageVM.cs:
```diff
@@ -1,10 +1,82 @@
+        private readonly IMarkService markService;
+        private readonly string username;
+        private List<Rating> marks;
+        private List<Rating> currentPageMarks;
```
Probably, we should convert this field to internal value field of method?

RomanLevkovych commented on ViewModel/ProfilePageVM.cs:
```diff
@@ -1,10 +1,82 @@
+            if (currentPageNumber < page_count-1)
+                currentPageMarks = marks.GetRange(currentPageNumber * pageLimit, (currentPageNumber + 1) * pageLimit - currentPageNumber * pageLimit);
+            if (currentPageNumber == page_count-1)
+                currentPageMarks = marks.GetRange(currentPageNumber * pageLimit, marks.Count % pageLimit);
```
Remove these ifs. Introduce variable for end of range and init it from ternary operator

ungidrid commented on DataServices/SqliteDbContext.cs:
```diff
@@ -16,7 +16,7 @@ public SqliteDbContext(string connectionString = null)
             var connectionStringBuilder = new SqliteConnectionStringBuilder();
             // TODO: create adequate location for DB.
-            connectionStringBuilder.DataSource = @"..\..\..\..\DataServices\univerity-db.db";
+            connectionStringBuilder.DataSource = @"..\..\..\..\DataServices\university-db.db";
```
Use normal sql connection string, not path to do file

ungidrid commented on View/Pages/ProfilePage.xaml.cs:
```diff
@@ -20,34 +24,154 @@ namespace View.Pages
+        private SortMarkTable sorted;
+        private readonly ProfilePageVM logic;
+        private readonly int pageLimit = 4;
+        int current_page_number = 0;
```
Use camelCase

ungidrid commented on View/Pages/ProfilePage.xaml.cs:
```diff
@@ -20,34 +24,154 @@ namespace View.Pages
+            if (sorted == SortMarkTable.Mark)
+            {
+                SortMarksTableRows(SortMarkTable.MarkReverse);
+                sorted = SortMarkTable.MarkReverse;
```
Put enum value into local variable using ternary operator so to make your code DRY

OlegAndrus12 commented on DataServices/DataServiceBase.cs:
```diff
@@ -111,6 +111,21 @@ public async Task<IList<Subject>> LoadSubjectsAsync()
 
+        public Student LoadStudent(Person person)
+        {
+            return _dataSource.Student.SingleOrDefault(p => p.Person.Username == person.Username);
```
Hmhmh...
Thanks, I`ll take it into consideration.
Den, it a joke or you really have to much free time to review other teams code?

RomanLevkovych replied:
I asked for this

OlegAndrus12 committed:
Code refactoring

Code refactored according to Den`s and Roman`s comments.
Sort column by marks is fixed

OlegAndrus12 committed:
Path to db changed

OlegAndrus12 commented:
1) Normal connection string is the thing we are struggled with. Even in current version of development we do not use conString correctly. 
2) Do you know better way to save image, excluding BitmapImage?
3) How sholud looks path for image in db to work on any machine?
4) Other thing changed according to yours prepositions.

Thanks)

OlegAndrus12 committed:
Problem With Path Resolved

OlegAndrus12 commented:
I resolved problem with path using Directory.GetCurrenrDeirectory().
I think we can use this approach to conString in db.
Images in db shoud be stored in format Widgets\Images\photoName.png. (default falue in db for profile picture "Widgets\Images\profile_placeholder.jpg")
